File: src/main/rs/emulate/legacy/widget/type/ModelWidget.java
package rs.emulate.legacy.widget.type;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.OptionalInt;

import rs.emulate.legacy.widget.WidgetGroup;
import rs.emulate.legacy.widget.WidgetOption;
import rs.emulate.legacy.widget.script.LegacyClientScript;
import rs.emulate.shared.util.DataBuffer;

/**
 * A {@link WidgetGroup#MODEL} {@link Widget}.
 *
 * @author Major
 */
public final class ModelWidget extends Widget {

	/**
	 * A model used as part of a ModelWidget.
	 */
	public static final class Model {

		/**
		 * The id of the animation.
		 */
		private final OptionalInt animation;

		/**
		 * The id of the model.
		 */
		private final OptionalInt model;

		/**
		 * Creates the Media.
		 *
		 * @param model The model id, as an {@link OptionalInt}.
		 * @param animation The animation id, as an OptionalInt.
		 */
		public Model(OptionalInt model, OptionalInt animation) {
			this.model = model;
			this.animation = animation;
		}

		/**
		 * Gets the id of the animation, as an {@link OptionalInt}.
		 *
		 * @return The animation id.
		 */
		public OptionalInt getAnimation() {
			return animation;
		}

		/**
		 * Gets the id of the model, as an {@link OptionalInt}.
		 *
		 * @return The model id.
		 */
		public OptionalInt getId() {
			return model;
		}

	}

	/**
	 * The pitch applied to the model.
	 */
	private int pitch;

	/**
	 * The primary Model.
	 */
	private Model primary;

	/**
	 * The roll applied to the model.
	 */
	private int roll;

	/**
	 * The scale applied to the model.
	 */
	private int scale;

	/**
	 * The secondary Model.
	 */
	private Model secondary;

	/**
	 * Creates the ModelWidget.
	 *
	 * @param id The id of the ModelWidget.
	 * @param parent The parent id of the ModelWidget, as an {@link OptionalInt}. Must not be {@code null}.
	 * @param optionType The {@link WidgetOption} of the ModelWidget. Must not be {@code null}.
	 * @param content The content type of the ModelWidget.
	 * @param width The width of the ModelWidget, in pixels.
	 * @param height The width of the ModelWidget, in pixels.
	 * @param alpha The alpha of the ModelWidget, in pixels.
	 * @param hover The hover id of the ModelWidget, as an {@link OptionalInt}. Must not be {@code null}.
	 * @param scripts The {@link List} of {@link LegacyClientScript}s. Must not be {@code null}.
	 * @param option The {@link Option} of the ModelWidget, as an {@link Optional}. Must not be {@code null}.
	 * @param hoverText The hover text of the ModelWidget, as an {@link Optional}. Must not be {@code null}.
	 * @param defaultModel The default {@link Model}.
	 * @param secondaryModel The secondary {@link Model}.
	 * @param scale The scale of the model.
	 * @param pitch The pitch of the model.
	 * @param roll The roll of the model.
	 */
	public ModelWidget(int id, OptionalInt parent, WidgetOption optionType, int content, int width, int height, int alpha,
			OptionalInt hover, List<LegacyClientScript> scripts, Optional<Option> option, Optional<String> hoverText,
			Model defaultModel, Model secondaryModel, int scale, int pitch, int roll) {
		super(id, parent, WidgetGroup.MODEL, optionType, content, width, height, alpha, hover, scripts, option, hoverText);

		this.primary = defaultModel;
		this.secondary = secondaryModel;
		this.scale = scale;
		this.pitch = pitch;
		this.roll = roll;
	}

	@Override
	protected DataBuffer encodeBespoke() {
		int size = 3 * Short.BYTES;
		for (Model model : Arrays.asList(primary, secondary)) {
			size += model.getId().isPresent() ? Short.BYTES : Byte.BYTES;
			size += model.getAnimation().isPresent() ? Short.BYTES : Byte.BYTES;
		}

		DataBuffer buffer = DataBuffer.allocate(size);
		for (OptionalInt optional : Arrays.asList(primary.getId(), secondary.getId(), primary.getAnimation(),
				secondary.getAnimation())) {
			if (optional.isPresent()) {
				int id = optional.getAsInt();
				buffer.putByte(id >> 8 + 1);
				buffer.putByte(id);
			} else {
				buffer.putBoolean(false);
			}
		}

		buffer.putShort(scale);
		buffer.putShort(pitch);
		buffer.putShort(roll);

		return buffer.flip().asReadOnlyBuffer();
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/widget/type/ModelWidget.kt
package rs.emulate.legacy.widget.type

import rs.emulate.legacy.widget.Widget
import rs.emulate.legacy.widget.WidgetGroup
import rs.emulate.legacy.widget.WidgetGroup.MODEL
import rs.emulate.legacy.widget.WidgetOption
import rs.emulate.legacy.widget.script.LegacyClientScript
import rs.emulate.legacy.widget.type.ModelWidget.Model
import rs.emulate.shared.util.putBoolean
import rs.emulate.shared.util.putByte
import java.nio.ByteBuffer
import java.util.Arrays

/**
 * A [WidgetGroup.MODEL] [Widget].
 *
 * @param id The id of the ModelWidget.
 * @param parent The parent id of the ModelWidget.
 * @param optionType The [WidgetOption] of the ModelWidget.
 * @param content The content type of the ModelWidget.
 * @param width The width of the ModelWidget, in pixels.
 * @param height The width of the ModelWidget, in pixels.
 * @param alpha The alpha of the ModelWidget, in pixels.
 * @param hover The hover id of the ModelWidget.
 * @param scripts The [List] of [LegacyClientScript]s.
 * @param option The [Option] of the ModelWidget.
 * @param hoverText The hover text of the ModelWidget.
 * @param primary The default [Model].
 * @param secondary The secondary [Model].
 * @param scale The scale of the model.
 * @param pitch The pitch of the model.
 * @param roll The roll of the model.
 */
class ModelWidget(
    id: Int, parent: Int?, optionType: WidgetOption, content: Int, width: Int, height: Int, alpha: Int,
    hover: Int?, scripts: List<LegacyClientScript>, option: Option?, hoverText: String?,
    private val primary: Model,
    private val secondary: Model,
    private val scale: Int,
    private val pitch: Int,
    private val roll: Int
) : Widget(id, parent, MODEL, optionType, content, width, height, alpha, hover, scripts, option, hoverText) {

    /**
     * A model used as part of a [ModelWidget].
     */
    data class Model(val id: Int?, val animation: Int?)

    override fun encodeBespoke(): ByteBuffer {
        var size = 3 * java.lang.Short.BYTES
        for (model in Arrays.asList(primary, secondary)) {
            size += if (model.id != null) java.lang.Short.BYTES else java.lang.Byte.BYTES
            size += if (model.animation != null) java.lang.Short.BYTES else java.lang.Byte.BYTES
        }

        val buffer = ByteBuffer.allocate(size)
        for (value in listOf(primary.id, secondary.id, primary.animation, secondary.animation)) {
            if (value != null) {
                buffer.putByte(value shr 8 + 1)
                buffer.putByte(value)
            } else {
                buffer.putBoolean(false)
            }
        }

        buffer.putShort(scale.toShort())
        buffer.putShort(pitch.toShort())
        buffer.putShort(roll.toShort())

        return buffer.apply { flip() }
    }

}
